src/array_slice.cpp

```diff
@@ -0,0 +1,452 @@
+  data.ninvmu = 0;
+  bool needs_permeability = false;
+  for (int i = 0; i < num_components; ++i)
+    if (components[i] == Permeability) { needs_permeability = true; break; }
+  if (needs_permeability) 
+    FOR_MAGNETIC_COMPONENTS(c) if (gv.has_field(c)) {
+      if (data.ninvmu == 3) abort("more than 3 field components??");
+      data.invmu_cs[data.ninvmu] = c;
+      data.invmu_ds[data.ninvmu] = component_direction(c);
+      ++data.ninvmu;
+    }
+ 
+  loop_in_chunks(get_array_slice_chunkloop, (void *) &data,
+		 where, Centered, true, true);
+
+  /***************************************************************/
+  /* repeatedly call sum_to_all to consolidate full array slice  */
+  /* on all cores                                                */
+  /***************************************************************/
+#define BUFSIZE 1<<16 // use 64k buffer
+  if (has_imag)
+   { cdouble buffer[BUFSIZE];
+     cdouble *slice = (cdouble *)vslice;
+     int offset=0, remaining=slice_size;
+     while(remaining!=0)
+      { 
+        int size = (remaining > BUFSIZE ? BUFSIZE : remaining);
+        sum_to_all(slice + offset, buffer, size);
+        memcpy(slice+offset, buffer, size*sizeof(cdouble));
+        remaining-=size;
+        offset+=size;
+      };
+   }
+  else
+   { double buffer[BUFSIZE];
+     double *slice = (double *)vslice;
+     int offset=0, remaining=slice_size;
+     while(remaining!=0)
+      { int size = (remaining > BUFSIZE ? BUFSIZE : remaining);
+        sum_to_all(slice + offset, buffer, size);
+        memcpy(slice+offset, buffer, size*sizeof(double));
+        remaining-=size;
+        offset+=size;
+      };
+   };
+
+  delete[] data.offsets;
+  delete[] data.fields;
+  delete[] data.ph;
+  delete[] data.cS;
+  finished_working();
+
+  return vslice;
+}
+
+/***************************************************************/
+/* entry points to get_array_slice                             */
+/***************************************************************/
+double *fields::get_array_slice(const volume &where,
+                                std::vector<component> components,
+                                field_function fun, void *fun_data,
+                                double *slice)
+{
+  return (double *)do_get_array_slice(where, components,
+                                      fun, fun_data, false,
+                                      (void *)slice);
+} 
+
+cdouble *fields::get_complex_array_slice(const volume &where,
+                                         std::vector<component> components,
+                                         field_function fun, void *fun_data,
+                                         cdouble *slice)
+{
+  return (cdouble *)do_get_array_slice(where, components,
+                                       fun, fun_data, true,
+                                       (void *)slice);
+}
+
+double *fields::get_array_slice(const volume &where, component c,
```

Comment:
It would also be nice to support `derived_component`, for common things like the energy density, similar to h5fields.